warn/scrapers/ga.py

```diff
@@ -1,121 +1,108 @@
-        )
-
-        # Concatenate the rows
-        output_rows.extend(new_rows)
+    def is_header(self):
+        """Test whether the row is a header."""
+        return self.item[0].tag == "th"
 
-    # Write out the results
-    data_path = data_dir / f"{state_code}.csv"
-    utils.write_rows_to_csv(data_path, output_rows)
+    def get_text(self):
+        """Get the text of the row."""
+        return tuple(utils.clean_text(cell.text_content()) for cell in self.item)
 
-    # Return the path to the CSV
-    return data_path
 
+class HmlTable(HtmlListPage):
+    """A table of HTML elements."""
 
-def _parse_table(html, id, include_headers=True):
-    """
-    Parse HTML table with given ID.
+    selector = CSS("table tr")
 
-    Keyword arguments:
-    html -- the HTML to parse
-    id -- the ID of the table to parse
-    include_headers -- whether to include the headers in the output (default True)
+    def process_item(self, item):
+        """Process an item."""
+        row = HtmlTableRow(item)
 
-    Returns: a list of rows
-    """
-    # Parse out data table
-    soup = BeautifulSoup(html, "html.parser")
-    table_list = soup.find_all(id=id)  # output is list-type
+        # Skip the header row, saving headers as property
+        if row.is_header():
+            self.headers = row
+            raise SkipItem("headers")
 
-    # We expect the first table to be there with our data
-    assert len(table_list) > 0
-    table = table_list[0]
+        return dict(zip(self.headers.get_text(), row.get_text()))
 
-    output_rows = []
-    column_tags = ["td"]
 
-    if include_headers:
-        column_tags.append("th")
+class NoticeList(HmlTable):
+    """The list of notices for a given year."""
 
-    # Loop through the table and grab the data
-    for table_row in table.find_all("tr"):
-        columns = table_row.find_all(column_tags)
-        output_row = []
+    selector = CSS("#emplrList tr")
 
-        for column in columns:
-            # Collapse newlines
-            partial = re.sub(r"\n", " ", column.text)
-            # Standardize whitespace
-            clean_text = re.sub(r"\s+", " ", partial).strip()
-            output_row.append(clean_text)
 
-        # Skip any empty rows
-        if len(output_row) == 0 or output_row == [""]:
-            continue
+def scrape(
+    data_dir: Path = utils.WARN_DATA_DIR,
+    cache_dir: Path = utils.WARN_CACHE_DIR,
+) -> Path:
+    """
+    Scrape data from Georgia.
 
-        output_rows.append(output_row)
+    Keyword arguments:
+    data_dir -- the Path were the result will be saved (default WARN_DATA_DIR)
+    cache_dir -- the Path where results can be cached (default WARN_CACHE_DIR)
 
-    return output_rows
+    Returns: the Path where the file is written
+    """
+    # Loop through the years and scrape them one by one
+    source = NullSource()
+    source.cache_dir = cache_dir / "ga"
+    years = YearList(start=2002, source=source)
```

Comment:
Serendipitous find! Turns out the first year listed on the interface is 2002 (because it lists 20 years by default), but the first year available appears to be 1989.